File: app/src/main/java/com/asfoundation/wallet/repository/GasSettingsRepository.java
package com.asfoundation.wallet.repository;

import androidx.annotation.NonNull;
import com.asfoundation.wallet.entity.GasSettings;
import io.reactivex.Observable;
import io.reactivex.Single;
import java.math.BigDecimal;
import java.util.concurrent.TimeUnit;
import org.web3j.protocol.Web3j;
import org.web3j.protocol.core.methods.response.EthGasPrice;

import static com.asfoundation.wallet.C.DEFAULT_GAS_LIMIT;
import static com.asfoundation.wallet.C.DEFAULT_GAS_LIMIT_FOR_TOKENS;
import static com.asfoundation.wallet.C.DEFAULT_GAS_PRICE;

public class GasSettingsRepository implements GasSettingsRepositoryType {

  private final static long FETCH_GAS_PRICE_INTERVAL = 60;
  private final Web3jProvider web3jProvider;
  private BigDecimal cachedGasPrice;

  public GasSettingsRepository(Web3jProvider web3jProvider) {
    this.web3jProvider = web3jProvider;

    cachedGasPrice = new BigDecimal(DEFAULT_GAS_PRICE);
    Observable.interval(0, FETCH_GAS_PRICE_INTERVAL, TimeUnit.SECONDS)
        .doOnNext(l -> updateGasSettings())
        .subscribe(l -> {
        }, t -> {
        });
  }

  private void updateGasSettings() {
    final Web3j web3j = web3jProvider.get();
    try {
      EthGasPrice price = web3j.ethGasPrice()
          .send();
      cachedGasPrice = new BigDecimal(price.getGasPrice());
    } catch (Exception ex) { /* Quietly */ }
  }

  public Single<GasSettings> getGasSettings(boolean forTokenTransfer) {
    return Single.fromCallable(() -> {
      BigDecimal gasLimit = getGasLimit(forTokenTransfer);
      if (cachedGasPrice == null) {
        updateGasSettings();
      }
      return new GasSettings(cachedGasPrice, gasLimit);
    });
  }

  @NonNull private BigDecimal getGasLimit(boolean forTokenTransfer) {
    return forTokenTransfer ? new BigDecimal(DEFAULT_GAS_LIMIT_FOR_TOKENS)
        : new BigDecimal(DEFAULT_GAS_LIMIT);
  }
}

File: app/src/main/java/com/asfoundation/wallet/repository/GasSettingsRepository.kt
package com.asfoundation.wallet.repository

import com.appcoins.wallet.core.network.backend.api.GasService
import com.asfoundation.wallet.entity.GasSettings
import io.reactivex.Single
import it.czerwinski.android.hilt.annotations.BoundTo
import java.math.BigDecimal
import java.util.concurrent.TimeUnit
import javax.inject.Inject

@BoundTo(supertype = GasSettingsRepositoryType::class)
class GasSettingsRepository @Inject constructor(private val gasService: GasService) :
    GasSettingsRepositoryType {

  private var lastFlushTime = 0L
  private var cachedGasPrice: BigDecimal? = null

  override fun getGasSettings(forTokenTransfer: Boolean): Single<GasSettings> {
    return getGasPrice()
        .map { GasSettings(it, getGasLimit(forTokenTransfer)) }
  }

  private fun getGasPriceNetwork(): Single<BigDecimal> {
    return gasService.getGasPrice()
        .map { BigDecimal(it.price) }
        .doOnSuccess {
          cachedGasPrice = it
          lastFlushTime = System.nanoTime()
        }
        .onErrorReturn {
          if (cachedGasPrice == null) {
            BigDecimal(DEFAULT_GAS_PRICE)
          } else {
            cachedGasPrice
          }
        }
  }

  private fun shouldRefresh() =
      System.nanoTime() - lastFlushTime >= TimeUnit.MINUTES.toNanos(1) || cachedGasPrice == null

  private fun getGasPriceLocal(): Single<BigDecimal> = Single.just(cachedGasPrice)

  private fun getGasPrice(): Single<BigDecimal> {
    return if (shouldRefresh()) {
      getGasPriceNetwork()
    } else {
      getGasPriceLocal()
    }
  }

  private fun getGasLimit(forTokenTransfer: Boolean): BigDecimal {
    return if (forTokenTransfer) {
      BigDecimal(DEFAULT_GAS_LIMIT_FOR_TOKENS)
    } else {
      BigDecimal(DEFAULT_GAS_LIMIT)
    }
  }

  companion object {
    const val DEFAULT_GAS_LIMIT = "90000"
    const val DEFAULT_GAS_LIMIT_FOR_TOKENS = "144000"
    const val DEFAULT_GAS_PRICE = "30000000000"
  }

}